### PR TITLE
Allow to use a const in the initializer of a const enum

### PR DESCRIPTION
**Pull request by andy-ms**
Was there any reason we didn't do this already? It doesn't seem to cause errors in the presence of circularity.

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
**file1.ts**
```ts
import { b } from './file2';

export const enum X {
  a = 1,
  c = b
}
```

**file2**
```ts
export const b = 10;
```

In your branch this compiles cleanly under `--isolatedModules` but `file1.ts` will generate different (and possibly non-working) code under `transpileModule`

**Comment by andy-ms** [comment]
From discussion with @RyanCavanaugh: It looks like `--isolatedModules` just ignores `const` anyway and produces accesses like `X.c` everywhere, so this won't affect that.
Ping @ahejlsberg to see if there are any other reasons we don't support this.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -21859,13 +21859,19 @@ namespace ts {
...
+                        }
+                        const fromEnum = evaluateEnumMember(id, getSymbolOfNode(member.parent), id.escapedText);
+                        return fromEnum !== undefined ? fromEnum : getFromLiteralType(id);
+                    }
                     case SyntaxKind.ElementAccessExpression:
                     case SyntaxKind.PropertyAccessExpression:
                         const ex = <PropertyAccessExpression | ElementAccessExpression>expr;
                         if (isConstantMemberAccess(ex)) {
-                            const type = getTypeOfExpression(ex.expression);
+                            const type = checkExpression(ex.expression);
```
Comment (sandersn) [inline src/compiler/checker.ts]: why can't we use getTypeOfExpression anymore?

File: tests/baselines/reference/enumInitializers_const_circular.errors.txt [hunk 2]
```diff
@@ -0,0 +1,21 @@
+tests/cases/compiler/enumInitializers_const_circular.ts(3,9): error TS2474: In 'const' enum declarations member initializer must be constant expression.
+tests/cases/compiler/enumInitializers_const_circular.ts(10,9): error TS2474: In 'const' enum declarations member initializer must be constant expression.
+
+
+==== tests/cases/compiler/enumInitializers_const_circular.ts (2 errors) ====
+    const x = E.y;
+    const enum E {
+        y = x,
+            ~
+!!! error TS2474: In 'const' enum declarations member initializer must be constant expression.
```
Comment (sandersn) [inline tests/baselines/reference/enumInitializers_const_circular.errors.txt]: could do with a better error message here too; `x` is const, but circular.

**Comment by ajafff** [comment]
This only works for constants with literal types? Otherwise there's no way the value can be inlined.
There should probably be a test for non-literal types, e.g. `: string` or `: 1 | 2`

File: src/compiler/checker.ts (hunk 1 above)
Comment (andy-ms) [reply src/compiler/checker.ts]: We can, but it just has special handling for `CallExpression` that we don't need, then calls `checkExpression`.

File: tests/baselines/reference/constEnumErrors.errors.txt [hunk 3]
```diff
@@ -36,9 +38,13 @@ tests/cases/compiler/constEnumErrors.ts(42,9): error TS2478: 'const' enum member
         Y = E1.Z,
             ~~~~
 !!! error TS2474: In 'const' enum declarations member initializer must be constant expression.
```
[the remark this answers is not in the document]
Comment (andy-ms) [reply tests/baselines/reference/constEnumErrors.errors.txt]: Is it good enough to have the new error message?

File: src/compiler/checker.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/checker.ts]: oh, that makes sense. I thought `getTypeOfExpression` was the fast path that was called by `checkExpression`, but I guess it actually just has some initial fast-path checks before delegating to `checkExpression`.

File: tests/baselines/reference/constEnumErrors.errors.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (sandersn) [reply tests/baselines/reference/constEnumErrors.errors.txt]: Sure, I guess so.

**Comment by andy-ms** [comment]
@ahejlsberg Good to go?

**Comment by andy-ms** [comment]
@ahejlsberg bump

**Comment by andy-ms** [comment]
@ahejlsberg

**Comment by andy-ms** [comment]
@ahejlsberg

**Comment by mhegazy** [comment]
@andy-ms can you talk to @ahejlsberg about this change

**Pull request closed without merging**